### PR TITLE
framework(engine): add StopImpl callback in master impl

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve?

Issue Number: ref #6749

### What is changed and how it works?

This PR implements the normal path of job canceling, and adds `StopImpl` in job master callback

#### Review guide

- changes in `engine/framework/internal/eventloop`, add `Stop` interface to task
- changes in `engine/framework`, add `StopImpl`interface to MasterImpl.
- changes in `engine/pkg/orm`, support orm for new `job_ops` table.
- `engine/servermaster/jobop/operator.go`, a simple job operator, it provides `MarkJobCanceling` API and `Tick` entrance.

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/servermaster/jobop/operator.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+func (oper *JobOperatorImpl) MarkJobCanceled(ctx context.Context, jobID string) error {
+	return oper.setJobOperate(ctx, jobID, ormModel.JobOpStatusCanceled)
+}
+
+// Tick implements JobOperator.Tick
+func (oper *JobOperatorImpl) Tick(ctx context.Context) error {
+	if !oper.isRunning.CAS(false, true) {
+		return nil
+	}
+	defer oper.isRunning.Swap(true)
```
Comment (liuzix) [inline engine/servermaster/jobop/operator.go:89]: Why is this necessary?

File: engine/servermaster/jobop/operator.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+// NewJobOperatorImpl creates a new JobOperatorImpl
+func NewJobOperatorImpl(cli pkgOrm.Client, router operateRouter) *JobOperatorImpl {
+	return &JobOperatorImpl{
+		frameMetaClient: cli,
+		router:          router,
+		isRunning:       *atomic.NewBool(false),
+	}
+}
+
+func (oper *JobOperatorImpl) setJobOperate(
```
Comment (liuzix) [inline engine/servermaster/jobop/operator.go:54]: Is `updateJobOperationStatus` a better name?

File: engine/servermaster/jobmanager.go [hunk 3]
```diff
@@ -459,6 +475,11 @@ func (jm *JobManagerImpl) Tick(ctx context.Context) error {
 		return err
 	}
 
+	// FIXME: Is this too frequent?
+	if err := jm.jobOperator.Tick(ctx); err != nil {
```
Comment (liuzix) [inline engine/servermaster/jobmanager.go:479]: Is it feasible to let `jobOperator` maintain an internal goroutine?

File: engine/pkg/orm/client.go [hunk 4]
```diff
@@ -641,6 +651,73 @@ func (c *metaOpsClient) GetOneResourceForGC(ctx context.Context) (*resModel.Reso
...
+			return errors.WrapError(errors.ErrMetaOpFail, err)
+		}
+		result.rowsAffected = exec.RowsAffected
+		return nil
+	})
+	return result, errors.WrapError(errors.ErrMetaOpFail, err)
+}
+
+// SetJobCanceled sets a cancelled status if a cancelling op exists.
+// If job op status is not found, does nothing.
```
Comment (liuzix) [inline engine/pkg/orm/client.go:690]: Can you comment on the behavior `If job op status is not found, does nothing`?

File: engine/servermaster/jobop/operator.go (hunk 1 above)
Comment (amyangfei) [reply engine/servermaster/jobop/operator.go]: removed this logic. I was expect `Tick` can be called concurrently, but current it doesn't need

File: engine/framework/base_jobmaster.go [hunk 5]
```diff
@@ -367,8 +377,24 @@ func (j *jobMasterImplAsWorkerImpl) Workload() model.RescUnit {
...
-	return j.inner.OnJobManagerMessage(topic, message)
+func (j *jobMasterImplAsWorkerImpl) OnMasterMessage(
+	ctx context.Context, topic p2p.Topic, message interface{},
+) error {
+	switch msg := message.(type) {
+	case *frameModel.StatusChangeRequest:
+		switch msg.ExpectState {
+		case frameModel.WorkerStatusStopped:
+			if !j.onCancelCalled {
+				j.onCancelCalled = true
```
Comment (maxshuang) [inline engine/framework/base_jobmaster.go:388]: We can add a `Info` log here to observe the framework `Cancel` behavior.

File: engine/framework/internal/master/worker_manager.go [hunk 6]
```diff
@@ -508,7 +508,7 @@ func (m *WorkerManager) checkWorkerEntriesOnce() error {
 			case frameModel.WorkerStatusFinished:
 				offlineError = derror.ErrWorkerFinish.FastGenByArgs()
 			case frameModel.WorkerStatusStopped:
```
Comment (maxshuang) [inline engine/framework/internal/master/worker_manager.go:510]: will we unify the `status` in other pr? Like:
`WorkerStatusStopped` -> `WorkerStatusCanceled`

File: engine/framework/master.go [hunk 7]
```diff
@@ -93,6 +94,9 @@ type MasterImpl interface {
 
 	// CloseImpl is called when the master is being closed
 	CloseImpl(ctx context.Context) error
+
+	// StopImpl is called when the master is being canceled
```
Comment (maxshuang) [inline engine/framework/master.go:98]: I think the comment for `CloseImpl` and `StopImpl` is inaccurate.
* `CloseImpl` is called when the job is not the terminated state;
* `StopImpl` is called when the job is  the terminated state;
terminated state is finished/failed/canceled.

Comment (amyangfei) [reply engine/framework/master.go:98]: I think `StopImpl` is called when job is failed or canceled, `finished` job should use `CloseImpl` logic.

File: engine/pkg/orm/model/jobop.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	"op",
+	"job_id",
+}
+
+// JobOp stores job operation recoreds
+type JobOp struct {
+	Model
+	Op    JobOpStatus `gorm:"type:tinyint not null;index:idx_job_op,priority:2;comment:Canceling(1),Canceled(2)"`
+	JobID string      `gorm:"type:varchar(128) not null;uniqueIndex:uk_job_id"`
+}
```
Comment (maxshuang) [inline engine/pkg/orm/model/jobop.go:46]: What about adding `Op` to job meta table directly?
What the motivation to use an isolated table to store the `canceling` action?

Comment (amyangfei) [reply engine/pkg/orm/model/jobop.go:46]: Do you mean table `master_meta_kv_data`?
To use an independent table to avoid data race, since both job manager and job master could update rows in table `master_meta_kv_data`

Comment (amyangfei) [reply engine/pkg/orm/model/jobop.go:46]: As we have discussed before, cancel job is a high level concept, which is not needed to store in master meta data.

Comment (maxshuang) [reply engine/pkg/orm/model/jobop.go:46]: Reasonable.

File: engine/framework/internal/master/worker_manager.go (hunk 6 above)
Comment (amyangfei) [reply engine/framework/internal/master/worker_manager.go:510]: yes

File: engine/servermaster/jobmanager.go [hunk 9]
```diff
@@ -613,3 +646,33 @@ func (jm *JobManagerImpl) WatchJobStatuses(
...
+		timer := time.NewTimer(jobOperateInterval)
+		for {
+			select {
+			case <-ctx.Done():
+				return errors.Trace(ctx.Err())
+			case <-jm.jobOperatorNotifier:
+			case <-timer.C:
+			}
+			if err := jm.jobOperator.Tick(ctx); err != nil {
+				inErr := jm.Exit(ctx, framework.ExitReasonFailed, err, "")
```
Comment (maxshuang) [inline engine/servermaster/jobmanager.go:661]: Why we expect the whole jobmanager exit when job operator is fail?
What about keep log error and try?

**Comment by maxshuang** [comment]
I will review it carefully later.

File: engine/framework/internal/eventloop/runner.go [hunk 10]
```diff
@@ -62,6 +62,14 @@ func (r *Runner[R]) Run(ctx context.Context) error {
 		panic(fmt.Sprintf("unexpected exiting with nil error: %s", r.task.ID()))
 	}
 
+	if isCanceledError(err) {
+		if stopErr := r.task.Stop(ctx); stopErr != nil {
+			log.Warn("Cancel task returned error",
+				zap.String("label", r.task.ID()), zap.Error(stopErr))
+		}
+		return err
```
Comment (maxshuang) [inline engine/framework/internal/eventloop/runner.go:70]: Why doesn't `CanceledError`  trigger `GracefulExit`?

File: engine/servermaster/jobmanager.go [hunk 11]
```diff
@@ -84,41 +90,57 @@ type JobManagerImpl struct {
...
-	if job == nil {
-		// Check if the job is not found.
-		if _, err := jm.frameMetaClient.GetJobByID(ctx, req.Id); pkgOrm.IsNotFoundError(err) {
+	meta, err := jm.frameMetaClient.GetJobByID(ctx, req.Id)
+	if err != nil {
+		if pkgOrm.IsNotFoundError(err) {
 			return nil, ErrJobNotFound.GenWithStack(&JobNotFoundError{JobID: req.Id})
 		}
-		return nil, ErrJobNotRunning.GenWithStack(&JobNotRunningError{JobID: req.Id})
+		return nil, err
```
Comment (maxshuang) [inline engine/servermaster/jobmanager.go:103]: Seems that this error should be wrap as the grpc error by status.Error()

File: engine/servermaster/jobmanager.go [hunk 12]
```diff
@@ -84,41 +90,57 @@ type JobManagerImpl struct {
...
+	pbJob, err := buildPBJob(meta)
+	if err != nil {
+		return nil, err
+	}
+	if pbJob.Status == pb.Job_Finished || pbJob.Status == pb.Job_Canceled {
+		return pbJob, nil
 	}
+
+	if err := jm.jobOperator.MarkJobCanceling(ctx, req.Id); err != nil {
+		return nil, err
```
Comment (maxshuang) [inline engine/servermaster/jobmanager.go:115]: ditto

File: engine/framework/master.go (hunk 7 above)
Comment (maxshuang) [reply engine/framework/master.go:98]: Have we agreed that `CloseImpl` is called `when the job is not the terminated state`?

File: engine/servermaster/jobmanager.go [hunk 13]
```diff
@@ -579,7 +607,12 @@ func (jm *JobManagerImpl) OnWorkerStatusUpdated(worker framework.WorkerHandle, n
 // CloseImpl implements frame.MasterImpl.CloseImpl
 func (jm *JobManagerImpl) CloseImpl(ctx context.Context) error {
 	jm.notifier.Close()
-	return nil
+	return jm.wg.Wait()
+}
+
+// StopImpl implements frame.MasterImpl.StopImpl
```
Comment (maxshuang) [inline engine/servermaster/jobmanager.go:613]: `bgJobOperatorLoop` may call `Exit` and trigger `StopImpl` for the  jobmananger.
Will it be a problem?

File: engine/servermaster/jobmanager.go (hunk 11 above)
Comment (amyangfei) [reply engine/servermaster/jobmanager.go:103]: If we don't wrap gRPC status, gRPC will automatically wrap it. Maybe we should add unified wrapper as a middleware instead of wrapping it everywhere. This PR won't do it.

```
plain error
apple@dataflow3:~/code/scripts » ./cancel_job.sh c3d79fcf-6e60-4b4d-9261-fcbb806dca62
{"code":2, "message":"test cancel job error"}%

normalized error
apple@dataflow3:~/code/scripts » ./cancel_job.sh c3d79fcf-6e60-4b4d-9261-fcbb806dca62
{"code":2, "message":"JobNotFoundError: {\"JobID\":\"c3d79fcf-6e60-4b4d-9261-fcbb806dca62\"}"}

rpc wrapped error
apple@dataflow3:~/code/scripts » ./cancel_job.sh c3d79fcf-6e60-4b4d-9261-fcbb806dca62
{"code":10, "message":"JobNotFoundError: {\"JobID\":\"c3d79fcf-6e60-4b4d-9261-fcbb806dca62\"}"}
```

File: engine/framework/internal/eventloop/runner.go (hunk 10 above)
Comment (amyangfei) [reply engine/framework/internal/eventloop/runner.go:70]: Because cancel is a force stop scenario, in which job master doesn't need to wait workers exit.

File: engine/servermaster/jobmanager.go (hunk 13 above)
Comment (amyangfei) [reply engine/servermaster/jobmanager.go:613]: Yes, changed the implement of `StopImpl` in job manager.

File: engine/servermaster/jobmanager.go (hunk 9 above)
Comment (amyangfei) [reply engine/servermaster/jobmanager.go]: addressed

File: engine/pkg/orm/client.go [hunk 14]
```diff
@@ -33,6 +33,8 @@ import (
 var globalModels = []interface{}{
 	&model.ProjectInfo{},
 	&model.ProjectOperation{},
+	&model.LogicEpoch{},
```
Comment (maxshuang) [inline engine/pkg/orm/client.go:36]:
```suggestion
```

File: engine/servermaster/jobop/operator.go [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+	meta, err := oper.frameMetaClient.GetJobByID(ctx, jobID)
+	if err != nil {
+		if pkgOrm.IsNotFoundError(err) {
+			log.Warn("found orphan job operation", zap.String("job-id", jobID))
+			return oper.MarkJobNoop(ctx, jobID)
+		}
+		return err
+	}
+	switch meta.StatusCode {
+	case frameworkModel.MasterStatusFinished, frameworkModel.MasterStatusStopped:
```
Comment (maxshuang) [inline engine/servermaster/jobop/operator.go:118]: If a job is marked `canceling` and fail at last, I think we can also mark the op `Canceled` since `failed` is a terminated state.

File: engine/framework/internal/eventloop/runner.go [hunk 16]
```diff
@@ -115,6 +122,20 @@ func isForcefulExitError(errIn error) bool {
 		return true
 	}
 
+	if derrors.ErrWorkerCancel.Equal(errIn) {
+		// In canceling scenario business logic promises that all workers have
+		// been canceled, which means we don't need a graceful exit.
+		return true
```
Comment (maxshuang) [inline engine/framework/internal/eventloop/runner.go:128]: I think it' s only for the `jobmaster`.
For `Worker`, things become subtle here. 
DM may hope to maintain the worker's status in `OnWorkerOffline`. If we skip `GracefulExit` here, 
DM job master need to rely on the `OnWorkerStatusUpdate` to update its worker's status.

**Comment by maxshuang** [comment]
RestLGTM.

File: engine/servermaster/jobop/operator.go (hunk 15 above)
Comment (amyangfei) [reply engine/servermaster/jobop/operator.go:118]: Add a comment, will be added in another PR.

File: engine/servermaster/jobop/operator.go [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+	ops, err := oper.frameMetaClient.QueryJobOpsByStatus(ctx, ormModel.JobOpStatusCanceling)
+	if err != nil {
+		return err
+	}
+	for _, op := range ops {
+		err := oper.router.SendCancelJobMessage(ctx, op.JobID)
+		if err != nil {
+			log.Warn("send cancel message to job master failed",
+				zap.String("job-id", op.JobID), zap.Error(err))
+			if err := oper.checkJobStatus(ctx, op.JobID); err != nil {
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply engine/servermaster/jobop/operator.go:100]: Not a good way to keep trying, it would cause dead loop.

File: engine/servermaster/jobmanager.go [hunk 18]
```diff
@@ -613,3 +649,31 @@ func (jm *JobManagerImpl) WatchJobStatuses(
...
+				return errors.Trace(ctx.Err())
+			case <-jm.jobOperatorNotifier:
+			case <-timer.C:
+			}
+			if err := jm.jobOperator.Tick(ctx); err != nil {
+				// error returns from Tick is only caused by metastore error, so
+				// only log it and retry later.
+				log.Warn("job operator tick with error", zap.Error(err))
+			}
+			timer.Reset(jobOperateInterval)
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply engine/servermaster/jobmanager.go:668]: It is acceptable, canceling has no fast-response requirement

File: engine/framework/internal/eventloop/runner.go (hunk 16 above)
Comment (amyangfei) [reply engine/framework/internal/eventloop/runner.go]: Move the logic to NotifyExit function of job master

File: engine/servermaster/jobmanager.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (maxshuang) [reply engine/servermaster/jobmanager.go:668]: `15s` seems too long for user waiting a job being canceled.
What's the `force kill` timeout for cancel in your design?

[the remark this answers is not in the document]
Comment (amyangfei) [reply engine/servermaster/jobmanager.go:668]: quite long, such as 10minutes

[the remark this answers is not in the document]
Comment (amyangfei) [reply engine/servermaster/jobmanager.go:668]: Canceling is not a high frequent behavior.

File: engine/servermaster/jobop/operator.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (maxshuang) [reply engine/servermaster/jobop/operator.go:100]: Sorry for the expression.  
`Keep trying`  here is `try all the ops as possible`, not `try the error op infinitely.`

File: engine/framework/internal/eventloop/runner.go (hunk 16 above)
Comment (maxshuang) [reply engine/framework/internal/eventloop/runner.go]: Ok. But in my mind, `Cancel` is just a normal operation like job finish.
`jobmaster` exiting directly has pros and cons:
* Pros: `jobmaster` can call `StopImpl` quickly
* Cons: `jobmanager` need to wait a timeout to update its jobmaster status

**Review by maxshuang — APPROVED** [review]

File: engine/servermaster/jobop/operator.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (amyangfei) [reply engine/servermaster/jobop/operator.go:102]: addressed

File: engine/servermaster/jobmanager.go [hunk 19]
```diff
@@ -613,3 +649,31 @@ func (jm *JobManagerImpl) WatchJobStatuses(
...
+
+func (jm *JobManagerImpl) bgJobOperatorLoop(ctx context.Context) {
+	jm.wg.Go(func() error {
+		timer := time.NewTimer(jobOperateInterval)
+		for {
+			select {
+			case <-ctx.Done():
+				return errors.Trace(ctx.Err())
+			case <-jm.jobOperatorNotifier:
+			case <-timer.C:
```
Comment (liuzix) [inline engine/servermaster/jobmanager.go:661]: Why do we need both `jm.jobOperatorNotifier` and a timer?

Comment (amyangfei) [reply engine/servermaster/jobmanager.go]: Changed to use `notify.Notifier`, which contains both manually trigger and periodic trigger

File: engine/pkg/orm/client.go [hunk 20]
```diff
@@ -641,6 +652,98 @@ func (c *metaOpsClient) GetOneResourceForGC(ctx context.Context) (*resModel.Reso
 	return &ret, nil
 }
 
+// SetJobNoop sets a job noop status if a job op record exists and status is canceling
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply engine/pkg/orm/client.go]: added

**Comment by amyangfei** [comment]
ptal @liuzix

**Review by liuzix — APPROVED** [review]
